File: Automower/Code/Navigation/姿态解算_m/eul2quatf.m
function q = eul2quatf(eul)
%#codegen

%Input:euler angle 3*1
%Output:quaternion 4*1

% Compute sines and cosines of half angles
c = cos(eul/2);
s = sin(eul/2);

% Construct quaternion
q = [c(1).*c(2).*c(3)+s(1).*s(2).*s(3);...
     c(1).*c(2).*s(3)-s(1).*s(2).*c(3);...
     c(1).*s(2).*c(3)+s(1).*c(2).*s(3);...
     s(1).*c(2).*c(3)-c(1).*s(2).*s(3)];
        
end

File: Automower/Code/Navigation/c_code_float_in_use/Att_ekf_related_function/eul2quatf.c
/* Include Files */
#include <math.h>
#include <stddef.h>
#include <stdlib.h>

/* Function Definitions */
/*
 * Input:euler angle 3*1
 * Output:quaternion 4*1
 *  Compute sines and cosines of half angles
 * Arguments    : const float eul[3]
 *                float q[4]
 * Return Type  : void
 */
void eul2quatf(float eul[3], float q[4])
{
  float c[3];
  float s[3];
  int k;

  /* 'eul2quatf:8' c = cos(eul/2); */
  /* 'eul2quatf:9' s = sin(eul/2); */
  for (k = 0; k < 3; k++) {
    c[k] = cos(eul[k] / 2.0);
    s[k] = sin(eul[k] / 2.0);
  }

  q[0] = c[0] * c[1] * c[2] + s[0] * s[1] * s[2];
  q[1] = c[0] * c[1] * s[2] - s[0] * s[1] * c[2];
  q[2] = c[0] * s[1] * c[2] + s[0] * c[1] * s[2];
  q[3] = s[0] * c[1] * c[2] - c[0] * s[1] * s[2];
}

